numpy/core/_add_newdocs.py

```diff
@@ -1350,6 +1350,27 @@
     >>> np.arange(3,7,2)
     array([3, 5])
 
+    .. WARNING::
+        Using non-integer steps in ``np.arange`` is discouraged and may lead to
+        confusing and non intuitive results.
+        For example, using ``dtype=np.int`` with fractional steps will cause
+        this behaviour:
+
+        >>> np.arange(-0.09, 10.0, 1.9, dtype=int)
+        array([0, 1, 2, 3, 4, 5])
+        >>> np.arange(-0.01, 10.0, 1.009, dtype=int)
+        array([0, 0, 0, 0, 0, 0, 0, 0, 0, 0])
+
+        When ``step`` is considerably smaller than ``start``, numerical
+        instability will occur. Use ``np.linspace`` instead:
+
+        >>> delta = pow(2.0, -50)
+        >>> np.set_printoptions(precision=50)
+        >>> np.arange(10, 10 + 5 * delta, delta)
+        array([10., 10., 10., 10.])
+        >>> np.linspace(10, 10 + 5 * delta, 5)
+        array([10.               , 10.               , 10.000000000000002,
+               10.000000000000004, 10.000000000000004])
```

Comment:
I think there are actually multiple things going on in this example which make it confusing. One clear way to illustrate the "numerical instability" is to show how it impacts the determination of the total number of values in the array:

```python
>>> delta = 1e-15
>>> np.arange(10, 10 + 5 * delta, delta).shape
(6,)
>>> np.linspace(10, 10 + 5 * delta, 5).shape
(5,)

>>> delta = 2**-50 # ~8.88e-16; which is < np.finfo(np.float64).resolution
>>> np.arange(10, 10 + 5 * delta, delta).shape
(4,)
>>> np.linspace(10, 10 + 5 * delta, 5).shape
(5,)
```
This example illustrates why linspace is desirable (because it always gives you `num` values).

The current example is tricky because the printed result returned by `linspace` is subject to the limitations of floating point precision. Conflating these two issues might be confusing for users.